libcnb-test/src/test_runner.rs

```diff
@@ -0,0 +1,216 @@
+    /// and/or Docker connection is required. For the common use-cases, use `Runner::default`.
+    pub fn new(tokio_runtime: tokio::runtime::Runtime, docker: Docker) -> Self {
+        TestRunner {
+            docker,
+            tokio_runtime,
+        }
+    }
+
+    /// Starts a new integration test run.
+    ///
+    /// This function will copy the application to a temporary directory, cross-compiles this crate,
+    /// packages it as a buildpack and then invokes [pack](https://buildpacks.io/docs/tools/pack/)
+    /// to build a new Docker image with the buildpacks specified by the passed [`TestConfig`].
+    ///
+    /// Since this function is supposed to only be used in integration tests, failures are not
+    /// signalled via [`Result`](Result) values. Instead, this function panics whenever an unexpected error
+    /// occurred to simplify testing code.
+    ///
+    /// # Panics
+    /// - When the app could not be copied
+    /// - When this crate could not be packaged as a buildpack
+    /// - When the `pack` command unexpectedly fails
+    ///
+    /// # Example
+    /// ```no_run
+    /// use libcnb_test::{assert_contains, TestRunner, TestConfig};
+    ///
+    /// TestRunner::default().run_test(
+    ///     TestConfig::new("heroku/builder:22", "test-fixtures/app"),
+    ///     |context| {
+    ///         assert_contains!(context.pack_stdout, "---> Ruby Buildpack");
+    ///         assert_contains!(context.pack_stdout, "---> Installing bundler");
+    ///         assert_contains!(context.pack_stdout, "---> Installing gems");
+    ///     },
+    /// )
+    /// ```
+    pub fn run_test<C: Borrow<TestConfig>, F: FnOnce(TestContext)>(&self, config: C, f: F) {
+        self.run_test_internal(util::random_docker_identifier(), config, f);
+    }
+
+    pub(crate) fn run_test_internal<C: Borrow<TestConfig>, F: FnOnce(TestContext)>(
+        &self,
+        image_name: String,
+        config: C,
+        f: F,
+    ) {
+        let config = config.borrow();
+
+        let app_dir = if config.app_dir.is_relative() {
+            env::var("CARGO_MANIFEST_DIR")
+                .map(PathBuf::from)
+                .expect("Could not determine Cargo manifest directory")
+                .join(&config.app_dir)
+        } else {
+            config.app_dir.clone()
+        };
+
+        let temp_app_dir =
+            app::copy_app(&app_dir).expect("Could not copy app to temporary location");
+
+        if let Some(app_dir_preprocessor) = &config.app_dir_preprocessor {
+            (app_dir_preprocessor)(PathBuf::from(temp_app_dir.path()));
+        }
+
+        let temp_crate_buildpack_dir = if config.buildpacks.contains(&BuildpackReference::Crate) {
+            Some(
+                build::package_crate_buildpack(&config.target_triple)
+                    .expect("Could not package current crate as buildpack"),
+            )
+        } else {
+            None
+        };
```

Comment:
(Optional) As of Rust 1.50+ we have [bool::then()](https://doc.rust-lang.org/std/primitive.bool.html#method.then). I can't decide whether I prefer it for this case or not? (Though wondered if you might leap at the chance to use a combinator 😆 😄)

```suggestion
        let temp_crate_buildpack_dir =
            config
                .buildpacks
                .contains(&BuildpackReference::Crate)
                .then(|| {
                    build::package_crate_buildpack(&config.target_triple)
                        .expect("Could not package current crate as buildpack")
                });
```

Reply:
🙌🏻  [45bda1f](https://github.com/heroku/libcnb.rs/pull/422/commits/45bda1f4d59d4a2fc5d83ea43ed5e0f55f806cf7)